synapse/storage/registration.py

```diff
@@ -208,14 +208,26 @@ def _user_delete_access_tokens(self, txn, user_id):
         )
 
     @defer.inlineCallbacks
-    def flush_user(self, user_id):
-        rows = yield self._execute(
-            'flush_user', None,
-            "SELECT token FROM access_tokens WHERE user_id = ?",
-            user_id
-        )
-        for r in rows:
-            self.get_user_by_access_token.invalidate((r,))
+    def user_delete_access_tokens_except(self, user_id, except_token_ids):
+        def f(txn):
+            txn.execute(
+                "SELECT id, token FROM access_tokens WHERE user_id = ? LIMIT 50",
+                    (user_id,)
+            )
```

Comment:
We don't really need to batch this `SELECT`, on the assumption that the number of tokens is likely to be less than the thousands. We tend to batch things to ensure that the SQL statement string itself isn't too long (as the length is limited), so its correct to batch the `DELETE` below.

The two options here are:
- Just batch the `DELETE` below and have the loop in the txn, its probably fine in this case since this function gets called relatively infrequently.
- Do the `SELECT` in a separate transaction and then do a transaction for each `DELETE` batch.